File: typescript/src/n0876_middle_of_the_linked_list.ts
import type { ListNode } from "./utils/linked_list.ts";

export function middleNode(head: ListNode | null): ListNode | null {
  let fast = head;
  let slow = head;

  while (fast?.next) {
    slow = slow!.next;
    fast = fast.next.next;
  }

  return slow;
}

File: python/src/n0876_middle_of_the_linked_list.py
from typing import Optional, cast

from utils.linked_list import ListNode


def middle_node(head: Optional[ListNode]) -> Optional[ListNode]:
    fast = head
    slow = head

    while fast and fast.next:
        fast = fast.next.next
        slow = cast(ListNode, slow).next

    return slow
